File: cpp/trex_selector_methods/trex_da/demo_trex_da_08_mc_sim_equi_and_bt/demo_trex_da_08_mc_sim_equi_and_bt.cpp
// ==============================================================================
// demo_trex_da_08_mc_sim_equi_and_bt.cpp
// ==============================================================================
/**
 * @file demo_trex_da_08_mc_sim_equi_and_bt.cpp
 *
 * @brief DA-TRex+EQUI Monte Carlo 2D sweep (SNR x rho) for the equicorrelated DGP.
 *
 * @details
 *
 *  One 2D sweep: for each SNR column in {0.5, 1, 2, 5}, run a full rho sweep
 *  over {0, 0.025, 0.05, 0.1, 0.2, 0.4, 0.7, 0.9} (dense near 0 to locate the
 *  DA suppression cliff) with the Equi DGP (n=300, p=1000, Random support),
 *  3 DA solvers + the 3 base T-Rex comparison rows.
 *
 *  Rationale: equicorrelation ties every predictor to one shared factor, so
 *  DA-EQUI suppresses all candidates as soon as rho is noticeably above 0,
 *  independent of signal strength, while base T-Rex loses FDR control at the
 *  same tiny rho. The 2D layout shows the cliff location, its SNR
 *  independence, and the rho=0 working anchor in a single figure.
 *
 *  Earlier revisions of this demo also carried BT (hierarchical-block) SNR
 *  and linkage sweeps; those were dropped as redundant — the BT method is
 *  covered on better-suited block designs in Demos 02-05.
 *
 * Results are saved to simulation_results/data/ as TXT and CSV files
 * (one file pair per SNR column).
 */
// ==============================================================================

#include "trex_da_sim_common.hpp"

// ==============================================================================

using namespace da_sim;

// ==============================================================================
//  2D sweep: SNR (outer, one output file per value) x rho (inner sweep)
// ==============================================================================

void demo_equi_mc_2d_snr_rho_sweep()
{
    constexpr int    n         = 300;
    constexpr int    p         = 1000;
    constexpr int    s         = 10;
    constexpr int    K         = 20;
    constexpr int    num_MC    = 200;
    constexpr double tFDR      = 0.2;
    constexpr int    base_seed = 2026;
    const auto       policy    = SupportPolicy::Random;

    const std::vector<double> snr_grid = {0.5, 1.0, 2.0, 5.0};
    const std::vector<double> rho_grid = {
        0.0, 0.025, 0.05, 0.1, 0.2, 0.4, 0.7, 0.9
    };

    TRexDAControlParameter da_ctrl;
    da_ctrl.method = DAMethod::EQUI;

    for (std::size_t si = 0; si < snr_grid.size(); ++si) {
        const double snr = snr_grid[si];

        // Filesystem-safe SNR tag: 0.5 -> "0p5", 2.0 -> "2".
        std::string snr_tag = fmt_num(snr);
        std::replace(snr_tag.begin(), snr_tag.end(), '.', 'p');

        // Disjoint seed blocks per SNR column (inner sweep spans
        // gi * 10000), so the columns are statistically independent.
        const int col_seed = base_seed + static_cast<int>(si) * 1000000;

        run_param_sweep(
            "da_equi_rho_snr" + snr_tag,
            "Rho",
            rho_grid,
            num_MC,
            tFDR,
            col_seed,
            default_solvers("+EQUI"),
            [&](double rho, unsigned seed) {
                return dgp_equi(n, p,
                                make_support(policy, s, p, 0, seed),
                                /*amplitude=*/3.0,
                                rho, snr, seed);
            },
            da_ctrl,
            make_trex_control(K),
            "Equi 2D sweep, SNR=" + fmt_num(snr)
            + ", n=" + std::to_string(n)
            + ", p=" + std::to_string(p)
            + ", support=" + support_policy_label(policy),
            /*include_base_trex=*/true);
    }
}


// ==============================================================================
//  main
// ==============================================================================

int main() {

    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);
    std::cout << "Running with " << omp_get_max_threads() << " threads\n\n";

    demo_equi_mc_2d_snr_rho_sweep();

    std::cout << "\nEqui 2D (SNR x rho) MC simulation complete.\n";
    return 0;
}

File: Python/trex_selector_methods/trex_da/demo_trex_da_08_mc_sim_equi_and_bt/demo_trex_da_08_mc_sim_equi_and_bt.py
# ==============================================================================
# demo_trex_da_08_mc_sim_equi_and_bt.py
# ==============================================================================
#
# DA-TRex+EQUI Monte Carlo 2D sweep (SNR x rho) for the equicorrelated DGP.
#
# One 2D sweep: for each SNR column in {0.5, 1, 2, 5}, run a full rho sweep
# over {0, 0.025, 0.05, 0.1, 0.2, 0.4, 0.7, 0.9} (dense near 0 to locate the
# DA suppression cliff) with the Equi DGP (n=300, p=1000, Random support),
# 3 DA solvers + the 3 base T-Rex comparison rows.
#
# Rationale: equicorrelation ties every predictor to one shared factor, so
# DA-EQUI suppresses all candidates as soon as rho is noticeably above 0,
# independent of signal strength, while base T-Rex loses FDR control at the
# same tiny rho. The 2D layout shows the cliff location, its SNR independence,
# and the rho=0 working anchor in a single figure.
#
# Earlier revisions of this demo also carried BT (hierarchical-block) SNR and
# linkage sweeps; those were dropped as redundant — the BT method is covered
# on better-suited block designs in Demos 02-05. (The folder keeps the cpp
# name "equi_and_bt" for parity.)
#
# Mirrors cpp/trex_selector_methods/trex_da/demo_trex_da_08_mc_sim_equi_and_bt
# (same DGP, control parameters, sweep axes, and result files — one TXT/CSV
# pair per SNR column). Results are saved to simulation_results/. Reduce
# NUM_MC to preview.
# ==============================================================================

import os
import sys

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

from da_sim_common import default_solvers, fmt_num, run_param_sweep

RUN_PART_1 = True

OUT_DIR = os.path.join(_THIS_DIR, "simulation_results")

# Simulation config (mirrors cpp)
N, P, S = 300, 1000, 10
AMPLITUDE = 3.0
TFDR = 0.2
K = 20
NUM_MC = 200
BASE_SEED = 2026

DA_SPEC = {"method": "EQUI"}


# ==============================================================================
# 2D sweep: SNR (outer, one output file per value) x rho (inner sweep)
# ==============================================================================

def part_1_2d_snr_rho_sweep():
    snr_grid = [0.5, 1.0, 2.0, 5.0]
    rho_grid = [0.0, 0.025, 0.05, 0.1, 0.2, 0.4, 0.7, 0.9]

    for si, snr in enumerate(snr_grid):
        # Filesystem-safe SNR tag: 0.5 -> "0p5", 2.0 -> "2".
        snr_tag = fmt_num(snr).replace(".", "p")

        # Disjoint seed blocks per SNR column (inner sweep spans gi * 10000),
        # so the columns are statistically independent.
        col_seed = BASE_SEED + si * 1000000

        run_param_sweep(
            OUT_DIR, f"da_equi_rho_snr{snr_tag}", "Rho", rho_grid, NUM_MC,
            TFDR, col_seed, default_solvers("+EQUI"),
            lambda rho, snr=snr: dict(n=N, p=P, dgp="dgp_equi_snr",
                                      kwargs=dict(amplitude=AMPLITUDE, rho=rho,
                                                  snr=snr),
                                      support=None, s=S),
            DA_SPEC, K,
            f"Equi 2D sweep, SNR={fmt_num(snr)}, n={N}, p={P}, support=Random",
            include_base_trex=True)


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    if RUN_PART_1:
        part_1_2d_snr_rho_sweep()
    print("\nEqui 2D (SNR x rho) MC simulation complete.")
